app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt

```diff
@@ -0,0 +1,124 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.gleanplumb
+
+import android.app.Notification
+import android.app.PendingIntent
+import android.content.Context
+import androidx.core.app.NotificationManagerCompat
+import androidx.work.ExistingPeriodicWorkPolicy
+import androidx.work.PeriodicWorkRequest
+import androidx.work.WorkManager
+import androidx.work.Worker
+import androidx.work.WorkerParameters
+import kotlinx.coroutines.DelicateCoroutinesApi
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import mozilla.components.support.base.ids.SharedIdsHelper
+import org.mozilla.fenix.ext.components
+import org.mozilla.fenix.nimbus.FxNimbus
+import org.mozilla.fenix.nimbus.MessageSurfaceId
+import org.mozilla.fenix.onboarding.MARKETING_CHANNEL_ID
+import org.mozilla.fenix.onboarding.MESSAGE_NOTIFICATION_ID
+import org.mozilla.fenix.utils.IntentUtils
+import org.mozilla.fenix.utils.createBaseNotification
+import java.util.concurrent.TimeUnit
+
+/**
+ * Background [Worker] that polls Nimbus for available [Message]s at a given interval.
+ * A [Notification] will be created using the configuration of the next highest priority [Message]
+ * if it has not already been displayed.
+ */
+class MessageNotificationWorker(
+    context: Context,
+    workerParameters: WorkerParameters,
+) : Worker(context, workerParameters) {
+
+    @OptIn(DelicateCoroutinesApi::class) // GlobalScope usage
+    override fun doWork(): Result {
+        GlobalScope.launch(Dispatchers.IO) {
+            val messagingStorage = applicationContext.components.analytics.messagingStorage
+            val messages = messagingStorage.getMessages()
+            val nextMessage =
+                messagingStorage.getNextMessage(MessageSurfaceId.NOTIFICATION, messages)
+
+            nextMessage?.apply {
+                if (metadata.displayCount == 0) {
```

Comment:
In the other refactoring issue (https://mozilla-hub.atlassian.net/browse/EXP-2994?) we may want to look at handling when a notification is dismissed or pressed

Reply:
Yes, once https://github.com/mozilla-mobile/fenix/pull/28565 lands this will be a simpler.